### PR TITLE
sql: set an "application name" for queries run using the InternalExecutor

### LINKED ISSUE #25331: sql: TestShowSessions failed under stress
SHA: https://github.com/cockroachdb/cockroach/commits/f86d47f55456e6bfa3072388896a7e3eb95224d4

Parameters:
```
TAGS=
GOFLAGS=
```

Stress build found a failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=642160&tab=buildLog

```
I180505 05:45:23.204282 151962 server/server.go:779  [n?] monitoring forward clock jumps based on server.clock.forward_jump_check_enabled
I180505 05:45:23.206175 151962 server/config.go:539  [n?] 1 storage engine initialized
I180505 05:45:23.206195 151962 server/config.go:542  [n?] RocksDB cache size: 128 MiB
I180505 05:45:23.206203 151962 server/config.go:542  [n?] store 0: in-memory, size 0 B
I180505 05:45:23.208732 151962 server/node.go:376  [n?] **** cluster 82a3c642-50bb-49d8-a41b-e1b38eb8d189 has been created
I180505 05:45:23.208753 151962 server/server.go:1349  [n?] **** add additional nodes by specifying --join=127.0.0.1:42223
I180505 05:45:23.209467 151962 storage/store.go:1369  [n1,s1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I180505 05:45:23.209516 151962 server/node.go:506  [n1] initialized store [n1,s1]: disk (capacity=512 MiB, available=512 MiB, used=0 B, logicalBytes=6.9 KiB), ranges=1, leases=0, writes=0.00, bytesPerReplica={p10=7043.00 p25=7043.00 p50=7043.00 p75=7043.00 p90=7043.00 pMax=7043.00}, writesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=0.00 pMax=0.00}
I180505 05:45:23.209545 151962 server/node.go:354  [n1] node ID 1 initialized
I180505 05:45:23.209591 151962 gossip/gossip.go:333  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:42223" > attrs:<> locality:<> ServerVersion:<major_val:2 minor_val:0 patch:0 unstable:3 > 
I180505 05:45:23.209751 151962 storage/stores.go:222  [n1] read 0 node addresses from persistent storage
I180505 05:45:23.209945 151962 server/node.go:647  [n1] connecting to gossip network to verify cluster ID...
I180505 05:45:23.210308 151962 server/node.go:672  [n1] node connected via gossip and verified as part of cluster "82a3c642-50bb-49d8-a41b-e1b38eb8d189"
I180505 05:45:23.210338 151962 server/node.go:440  [n1] node=1: started with [<no-attributes>=<in-mem>] engine(s) and attributes []
I180505 05:45:23.210427 151962 server/server.go:1474  [n1] starting https server at 127.0.0.1:45809
I180505 05:45:23.210442 151962 server/server.go:1475  [n1] starting grpc/postgres server at 127.0.0.1:42223
I180505 05:45:23.210455 151962 server/server.go:1476  [n1] advertising CockroachDB node at 127.0.0.1:42223
I180505 05:45:23.224432 151751 sql/event_log.go:124  [n1,intExec=optInToDiagnosticsStatReporting] Event: "set_cluster_setting", target: 0, info: {SettingName:diagnostics.reporting.enabled Value:true User:root}
I180505 05:45:23.247040 151758 sql/event_log.go:124  [n1,intExec=set-setting] Event: "set_cluster_setting", target: 0, info: {SettingName:version Value:$1 User:root}
I180505 05:45:23.253548 152567 sql/event_log.go:124  [n1,intExec=disableNetTrace] Event: "set_cluster_setting", target: 0, info: {SettingName:trace.debug.enable Value:false User:root}
I180505 05:45:23.271405 152577 sql/event_log.go:124  [n1,intExec=initializeClusterSecret] Event: "set_cluster_setting", target: 0, info: {SettingName:cluster.secret Value:gen_random_uuid()::STRING User:root}
I180505 05:45:23.274510 151962 server/server.go:1553  [n1] done ensuring all necessary migrations have run
I180505 05:45:23.274551 151962 server/server.go:1556  [n1] serving sql connections
I180505 05:45:23.276743 152856 sql/event_log.go:124  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:127.0.0.1:42223} Attrs: Locality: ServerVersion:2.0-3} ClusterID:82a3c642-50bb-49d8-a41b-e1b38eb8d189 StartedAt:1525499123210319928 LastUp:1525499123210319928}
I180505 05:45:23.307085 151962 server/server.go:779  [n?] monitoring forward clock jumps based on server.clock.forward_jump_check_enabled
I180505 05:45:23.309043 151962 server/config.go:539  [n?] 1 storage engine initialized
I180505 05:45:23.309068 151962 server/config.go:542  [n?] RocksDB cache size: 128 MiB
I180505 05:45:23.309079 151962 server/config.go:542  [n?] store 0: in-memory, size 0 B
W180505 05:45:23.309132 151962 gossip/gossip.go:1293  [n?] no incoming or outgoing connections
I180505 05:45:23.309167 151962 server/server.go:1351  [n?] no stores bootstrapped and --join flag specified, awaiting init command.
I180505 05:45:23.320360 152915 gossip/client.go:129  [n?] started gossip client to 127.0.0.1:42223
I180505 05:45:23.320642 152591 gossip/server.go:219  [n1] received initial cluster-verification connection from {tcp 127.0.0.1:40921}
I180505 05:45:23.321386 151962 storage/stores.go:222  [n?] read 0 node addresses from persistent storage
I180505 05:45:23.321422 151962 server/node.go:647  [n?] connecting to gossip network to verify cluster ID...
I180505 05:45:23.321442 151962 server/node.go:672  [n?] node connected via gossip and verified as part of cluster "82a3c642-50bb-49d8-a41b-e1b38eb8d189"
I180505 05:45:23.321910 152592 storage/stores.go:241  [n?] wrote 1 node addresses to persistent storage
I180505 05:45:23.323144 151962 kv/dist_sender.go:368  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
I180505 05:45:23.323545 151962 server/node.go:347  [n?] new node allocated ID 2
I180505 05:45:23.323587 151962 gossip/gossip.go:333  [n2] NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:40921" > attrs:<> locality:<> ServerVersion:<major_val:2 minor_val:0 patch:0 unstable:3 > 
I180505 05:45:23.323614 151962 server/node.go:421  [n2] node=2: asynchronously bootstrapping engine(s) [<no-attributes>=<in-mem>]
I180505 05:45:23.323623 151962 server/node.go:440  [n2] node=2: started with [] engine(s) and attributes []
I180505 05:45:23.323679 151962 server/server.go:1474  [n2] starting https server at 127.0.0.1:38713
I180505 05:45:23.323689 151962 server/server.go:1475  [n2] starting grpc/postgres server at 127.0.0.1:40921
I180505 05:45:23.323696 151962 server/server.go:1476  [n2] advertising CockroachDB node at 127.0.0.1:40921
I180505 05:45:23.332695 152906 storage/stores.go:241  [n1] wrote 1 node addresses to persistent storage
I180505 05:45:23.333214 151962 server/server.go:1553  [n2] done ensuring all necessary migrations have run
I180505 05:45:23.333234 151962 server/server.go:1556  [n2] serving sql connections
I180505 05:45:23.336280 152945 server/node.go:628  [n2] bootstrapped store [n2,s2]
I180505 05:45:23.384776 152840 sql/event_log.go:124  [n1,client=127.0.0.1:54170,user=root] Event: "create_database", target: 50, info: {DatabaseName:test Statement:CREATE DATABASE IF NOT EXISTS test User:root}
I180505 05:45:23.387922 152840 sql/event_log.go:124  [n1,client=127.0.0.1:54170,user=root] Event: "create_table", target: 51, info: {TableName:test.public.t Statement:CREATE TABLE test.public.t (num INT) User:root}
I180505 05:45:23.403531 153076 util/stop/stopper.go:471  quiescing; tasks left:
1      ts.poller: poll
W180505 05:45:23.403825 153032 server/node.go:835  [n2] node=2: unable to log node_join event: log-event: failed to send RPC: sending to all 1 replicas failed; last error: {<nil> rpc error: code = Unavailable desc = transport is closing}
W180505 05:45:23.403843 152951 ts/db.go:154  [n2,ts-poll] error writing time series data: failed to send RPC: sending to all 1 replicas failed; last error: {<nil> rpc error: code = Unavailable desc = transport is closing}
	show_test.go:625: unexpected number of running sessions: 2, expected 1
```

### PR DESCRIPTION
**Pull request by andreimatei**
Seems like a good idea in general, and it allows tests for `show
sessions` to filter them out.
TestShowSessions was flaky because it was running into unexpected
sessions created by various internal queries that run in the background.

Fixes #25331

Release note: None

### CODE REVIEW

File: pkg/sql/internal.go [hunk 1]
```diff
@@ -385,6 +386,9 @@ func (ie *internalExecutorImpl) execInternal(
 	stmt string,
 	qargs ...interface{},
 ) (retRes result, retErr error) {
+	if sargs != nil && sargs.ApplicationName == "" {
+		sargs.ApplicationName = fmt.Sprintf("internal-%s", opName)
```
Comment (knz) [inline pkg/sql/internal.go]: I'm not superplussed about this `Sprintf`.

The `opName` argument is a constant string in all the callers. It's silly (and inefficient, and slow, and expensive) to add the "internal" prefix *at run time* given that the strings are constant.

Instead, change the call points to add the "internal" prefix to the `opName` argument there, in the string constant.

**Review by knz — COMMENTED** [review]
Good call, but I'd do it differently. See comment below.

**Comment by jordanlewis** [comment]
Great! Thanks.

**Comment by andreimatei** [comment]
Shit, I realized I hadn't pushed before. Pushed now.

**Comment by knz** [comment]
lemme have another look then

**Comment by knz** [comment]
I am a bit concerned though about SHOW QUERIES. 

- if we make it ignore the special value, then there will be no easy way to cancel these queries.

- if we don't make it ignore the special values, stuff will show up in the output of SHOW QUERIES which doesn't have a matching session in SHOW SESSION

that seems like broken UX. 

What do you think? @jordanlewis any thoughts too?

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25673#-:-LDhJx8R1vn8WGDnIWli:bqoxshi)*

**Comment by knz** [comment]
(To be clear: until we resolve the UX issue, I'm overall :-1: on creating a special value for `application_name` that's ignored by SHOW QUERIES/SESSIONS).

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25673#-:-LDhLT3JE_sB45SpxC-7:b-9i1843)*

**Comment by andreimatei** [comment]
I'm not changing the behavior of `show sessions`. I'm only changing a test :)

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25673#-:-LDhPy2_9RSkWgrCYYyF:beda6s6)*

**Comment by andreimatei** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/internal.go, line 390 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25673#-LCooD2wS73XoNtB4rQ3-r1-390:-LDhR75SEREDBIERIxh0:b5lharu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef178593c998a3c31230f6c125758fd227d910b8/pkg/sql/internal.go#L390)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

> I really don't think the performance of this is important

Have you ever heard the phrase "death by a thousand cuts"? It's called Lingchi in Chinese: https://en.wikipedia.org/wiki/Lingchi

This is one cut.
</blockquote></details>

it's really not :)
Think of it this way - I'm saving space in the binary so I can put LLVM in it.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25673)*

**Comment by knz** [comment]
>  I'm only changing a test :)

Friend, that is obviously not true and you know it. You are making a user-visible change with an impact on UX. If you prefer to not care about this, that's OK by me, but then please hand this PR to a colleague (I could volunteer) so that I can complete the work. Right now the code as-is is not yet adequate, and I'd rather not rubber stamp this until we are happy about the quality.

**Comment by andreimatei** [comment]
I think there's a misunderstanding here - `show sessions` is not showing any more or any less sessions with this patch. It only shows some sessions which previously had no app name as having an "internal" app name. What's there not to like?

**Comment by knz** [comment]
hmm

**Comment by knz** [comment]
All right, I hadn't seen that this is only filtering the output of SHOW SESSIONS inside the test, and not anywhere else.

I am not awake today. My sincerest apologies for wasting your time.

LGTM

**Pull request merged**